init support for VxWorks-7 on Freescale i.MX6

yanyan-wrs opened the pull request:
Add changes to work with VxWorks as Rich OS:

1. added VxWorks specific SMC requests definition. OP-TEE is to be used as both a TEE and a Monitor for VxWorks, so additional SMC requests must be handled by OP-TEE.

2. added a new the hand-off protocol between loader and OP-TEE, that is, no hand-off at all. OP-TEE will not require loader to pass any arguments, instead, things such as paged data, NS entry and DT address are from configuration parameters. This hand-off is to be used by VxWorks.

3. Added the i.MX6 Sabre Lite board support.

jforissier commented:
@yanyan-wrs the "close" and "re-open" buttons are working, there is no need to test them so often ;)

yanyan-wrs commented:
@jforissier sorry, first time doing this. I thought I need to "close" the pull request to add new staffs. By the way, i noticed there has been no comment or action taken for this pull request, so, what's the holding up, am I doing something the wrong way?

jforissier commented:
@yanyan-wrs no problem ;)
The reason why this PR is not moving forward is because several people in our team are on vacation currently, so we are lacking reviewers. But don't worry we will definitely look at this as soon as possible.

etienne-lms commented on core/arch/arm/kernel/generic_entry_a32.S:
```diff
@@ -425,6 +445,9 @@ UNWIND(	.cantunwind)
 
 	bl	plat_cpu_reset_late
 
+    /* this core just comes out of reset, discard undefined l1 contents */
```
Minor issue of indentation of comment.

yanyan-wrs replied:
accepted and fixed

etienne-lms commented on core/arch/arm/kernel/generic_entry_a32.S:
```diff
@@ -425,6 +445,9 @@ UNWIND(	.cantunwind)
 	bl	plat_cpu_reset_late
 
+    /* this core just comes out of reset, discard undefined l1 contents */
+	bl	arm_cl1_d_invbysetway
```
Could you cross check this? If core is inside the smp coherency, could this affect other cores.

yanyan-wrs replied:
This is safe because its "cache invalidate" on L1, other cores would not respond to this op. Besides, the MMU and cache for secondary core is not enabled yet at this moment, so I believe there will not be any coherency broadcast even if the SMP bit is enable.

Anyway, my testing on i.MX6 (Cortex-A9) shows this operation is required for secondary cores to boot up. ARM specs also mentioned that when CPU comes out of reset, the cache content is UNDEFINED, so an all set-way invalidation is necessary.

And I wonder why this is not required before. Maybe this again is because the OP-TEE is targeted for Firmware; so it is possible that Firmware cleaned the cache before putting secondary cores into OP-TEE; however, for VxWorks's usage, the secondary core is released from reset and is directly put into OP-TEE, so the cache should be handled somehow here.

Anyway, I admit globally changing this is inconsiderate, I'll put it inside CFG_TEE_VXWORKS_SUPPORT to reflect this is due to VxWorks's special boot scenario.

etienne-lms replied:
> This is safe because its "cache invalidate" on L1, other cores would not
respond to this op.

ok. thanks for the details.
looks clear.


> And I wonder why this is not required (...)

on armv8, cache were cleaned by el3.
on armv7, it is likely that for those legacy platforms an earlier
bootloader invalidates caches.


> I'll put it inside CFG_TEE_VXWORKS_SUPPORT to reflect this is due to
VxWorks's special boot scenario.

since introducing CFG_TEE_VXWORKS_SUPPORT , it could be nice to have this
info inside optee.
Could you elaborate some make command using
$(call cfg-depends-one/all,... to detect the dependency, and integrate it
to mk/config.mk ?


On 12 August 2016 at 10:04, yanyan-wrs <notifications@github.com> wrote:

> In core/arch/arm/kernel/generic_entry_a32.S
> <https://github.com/OP-TEE/optee_os/pull/936#discussion_r74554282>:
>
> > @@ -425,6 +445,9 @@ UNWIND(	.cantunwind)
> >
> >  	bl	plat_cpu_reset_late
> >
> > +    /* this core just comes out of reset, discard undefined l1 contents */
> > +	bl	arm_cl1_d_invbysetway
>
> This is safe because its "cache invalidate" on L1, other cores would not
> respond to this op. Besides, the MMU and cache for secondary core is not
> enabled yet at this moment, so I believe there will not be any coherency
> broadcast even if the SMP bit is enable.
>
> Anyway, my testing on i.MX6 (Cortex-A9) shows this operation is required
> for secondary cores to boot up. ARM specs also mentioned that when CPU
> comes out of reset, the cache content is UNDEFINED, so an all set-way
> invalidation is necessary.
>
> And I wonder why this is not required before. Maybe this again is because
> the OP-TEE is targeted for Firmware; so it is possible that Firmware
> cleaned the cache before putting secondary cores into OP-TEE; however, for
> VxWorks's usage, the secondary core is released from reset and is directly
> put into OP-TEE, so the cache should be handled somehow here.
>
> Anyway, I admit globally changing this is inconsiderate, I'll put it
> inside CFG_TEE_VXWORKS_SUPPORT to reflect this is due to VxWorks's special
> boot scenario.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OP-TEE/optee_os/pull/936/files/94967ff72e6c157bd08e0134b1e82a5792cfd333#r74554282>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATEHJsT1eLSULwkjlwh4RAzH6Ngzz6Zmks5qfCkjgaJpZM4JTw0v>
> .
>

etienne-lms replied:
>  This is safe because its "cache invalidate" on L1, other cores would not respond to this op.

ok. thanks for the details.
looks clear.

> And I wonder why this is not required (...)

on armv8, cache were cleaned by el3.
on armv7, it is likely that for those legacy platforms an earlier bootloader invalidates caches.

> I'll put it inside CFG_TEE_VXWORKS_SUPPORT to reflect this is due to VxWorks's special boot scenario.

Since introducing CFG_TEE_VXWORKS_SUPPORT , it could be nice to have this dependency info within optee.
Could you elaborate some make command using `$(call cfg-depends-one/all,`... to detect the dependency, and integrate it into mk/config.mk ?

yanyan-wrs replied:
I refactored all changes into three independent CFG items:
CFG_TEE_VXWORKS_SMC: this enables VxWorks extended SMC handling in OP-TEE and corresponding plat-xxxx code.
CFG_TEE_BUILT_IN_ARGS: this enables OP-TEE to used built-in arguments instead of relying on arguments passed in from firmware.
CFG_TEE_SECONARY_ENTRY_L1_INV: this enables OP-TEE's generic boot process to invalidate L1 on secondary entry, necessary for platforms that do not have el3 firmware to take care of this.

Now there are just three independent CFGs and CFG_TEE_BUILT_IN_ARGS and CFG_TEE_SECONDARY_ENTRY_L1_INV can be used in cases other than VxWorks Support.

There is no long overall "VxWorks support" option, instead, users are required to enable VxWorks SMC Ext, Built-in ARGS and SMP L1 Inv to support VxWokrs. We'll document this in Wind River's User Guide.

All above options are covered in mk/config.mk.

etienne-lms commented on core/arch/arm/plat-imx6/imx6_init.S:
```diff
@@ -0,0 +1,345 @@
+ * It is assume each routine is called with return address in LR
+ * and with ARM registers R0, R1, R2, R3 being scratchable.
+ */
+#include <kernel/tz_ssvce_def.h>
```
Minor coding style issue: could alphabetically sort the include files. It is a good habbit in optee.

yanyan-wrs replied:
accepted and fixed

etienne-lms commented on core/arch/arm/kernel/generic_entry_a32.S:
```diff
@@ -434,6 +457,7 @@ UNWIND(	.cantunwind)
 	bl	cpu_is_ready
 
 	mov	r0, r5		/* ns-entry address */
+
```
Discard this change

yanyan-wrs replied:
accepted and fixed

etienne-lms commented on core/arch/arm/kernel/generic_entry_a32.S:
```diff
@@ -38,7 +38,14 @@
+ * secondary cores. Thus CPU Sync is not required.
+ */
+
+#if defined (CFG_BOOT_SYNC_CPU) && !defined (CFG_TEE_VXWORKS_SUPPORT)
```
You could force `CFG_BOOT_SYNC_CPU=n` in `plat-xxx/conf.mk`.

yanyan-wrs replied:
I understand we can set CFG_BOOT_SYNC_CPU=n in plat-xxx/conf.mk. However, the intention here is to "uniformly" disable the CPU sync feature if OP-TEE is to be used together with VxWorks. Besides, for some platform, for example, plat-ls, it already defines the CFG_BOOT_SYNC_CPU=y. We want to reuse the existing plat-ls to support VxWorks, but we cannot just rudely overwrite CFG_BOOT_SYNC_CPU=n, so we have to disable the feature using the CFG_TEE_VXWORKS_SUPPORT flag. 

It seems the generic boot process implementation is targeted for Firmware Code. The firmware would bring up all cores and put each through the OP-TEE; VxWorks's usage differs from this: only the primary core is required to go through the OP-TEE and enters NS OS (this maximize compatibility as it is not required to have TrustZone aware firmware), then NS OS asks the OP-TEE to bring up the rest cores, so there's no need to sync between primary and secondary cores.

etienne-lms replied:
> We want to reuse the existing plat-ls to support VxWorks

i would rather propose plat-stm and plat-ls to be changed:
-$(call force,CFG_BOOT_SYNC_CPU,y)
+CFG_BOOT_SYNC_CPU ?= y

To run VxWorks on a native optee/ls:
make CFG_BOOT_SYNC_CPU=n

Otherwise, in mk/config.mk
+ifeq ($(CFG_TEE_VXWORKS_SUPPORT),y)
+$(call force,CFG_BOOT_SYNC_CPU,n)
+endif

would that work?

My point is you intend to disable the full "boot sync cpu". So i would
rather have it disabled than adding extra check in the source code.


> It seems the generic boot process implementation is targeted for Firmware
Code.
> The firmware would bring up all cores and put each through the OP-TEE;
>  VxWorks's usage differs from this:

no so. ATF support on aarch32 expects configuration sets explicitly
CFG_WITH_ARM_TRUSTED_FW=y
CFG_BOOT_SYNC_CPU=n

CFG_BOOT_SYNC_CPU is designed for that purpose.
I think the real issue is plat-ls (as plat-stm) forcing cpu sync at boot.


On 12 August 2016 at 09:51, yanyan-wrs <notifications@github.com> wrote:

> In core/arch/arm/kernel/generic_entry_a32.S
> <https://github.com/OP-TEE/optee_os/pull/936#discussion_r74552988>:
>
> > @@ -38,7 +38,14 @@
> >  .section .data
> >  .balign 4
> >
> > -#ifdef CFG_BOOT_SYNC_CPU
> > +/*
> > + * For VxWorks Support, the primary core will go through OP-TEE init
> > + * without ever touching secondary cores; it is only after the primary
> > + * core enters VxWorks, it issues special SMC to request booting of
> > + * secondary cores. Thus CPU Sync is not required.
> > + */
> > +
> > +#if defined (CFG_BOOT_SYNC_CPU) && !defined (CFG_TEE_VXWORKS_SUPPORT)
>
> I understand we can set CFG_BOOT_SYNC_CPU=n in plat-xxx/conf.mk. However,
> the intention here is to "uniformly" disable the CPU sync feature if OP-TEE
> is to be used together with VxWorks. Besides, for some platform, for
> example, plat-ls, it already defines the CFG_BOOT_SYNC_CPU=y. We want to
> reuse the existing plat-ls to support VxWorks, but we cannot just rudely
> overwrite CFG_BOOT_SYNC_CPU=n, so we have to disable the feature using the
> CFG_TEE_VXWORKS_SUPPORT flag.
>
> It seems the generic boot process implementation is targeted for Firmware
> Code. The firmware would bring up all cores and put each through the
> OP-TEE; VxWorks's usage differs from this: only the primary core is
> required to go through the OP-TEE and enters NS OS (this maximize
> compatibility as it is not required to have TrustZone aware firmware), then
> NS OS asks the OP-TEE to bring up the rest cores, so there's no need to
> sync between primary and secondary cores.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OP-TEE/optee_os/pull/936/files/94967ff72e6c157bd08e0134b1e82a5792cfd333#r74552988>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATEHJt963UI4PbeVfCiQwQ5bnEweMhw1ks5qfCYBgaJpZM4JTw0v>
> .
>

etienne-lms replied:
> We want to reuse the existing plat-ls to support VxWorks

i would rather propose plat-stm and plat-ls to be changed:
```
-$(call force,CFG_BOOT_SYNC_CPU,y)
+CFG_BOOT_SYNC_CPU ?= y
```
To run VxWorks on a native optee/ls:
`make CFG_BOOT_SYNC_CPU=n`

Otherwise, in mk/config.mk
```
+ifeq ($(CFG_TEE_VXWORKS_SUPPORT),y)
+$(call force,CFG_BOOT_SYNC_CPU,n)
+endif
```

would that work?

My point is you intend to disable the full "boot sync cpu". So i would
rather have it disabled than adding extra check in the source code.


> It seems the generic boot process implementation is targeted for Firmware Code.
> The firmware would bring up all cores and put each through the OP-TEE;
>  VxWorks's usage differs from this:

no so. ATF support on aarch32 expects configuration sets explicitly
```
CFG_WITH_ARM_TRUSTED_FW=y
CFG_BOOT_SYNC_CPU=n
```
CFG_BOOT_SYNC_CPU is designed for that purpose.
I think the real issue is plat-ls (as plat-stm) forcing cpu sync at boot.

yanyan-wrs replied:
You're right it's better solution to modify the "conf.mk" of plat-ls and plat-stm not to force the CPU sync, and rather inject the CFG_BOOT_SYNC_CPU from make command. I prefer the "CFG_BOOT_SYNC_CPU ?= y" solution since it seems more flexible. I'll update the code accordingly.

etienne-lms commented on core/arch/arm/plat-imx6/platform_config.h:
```diff
@@ -0,0 +1,192 @@
+				CORE_MMU_DEVICE_SIZE)
+#define DEVICE0_VA_BASE         DEVICE0_PA_BASE
+#define DEVICE0_SIZE			CORE_MMU_DEVICE_SIZE
+#define DEVICE0_TYPE			MEM_AREA_IO_SEC
```
From comments inyour main.c, it seems the uart is shared you non secure.
Maybe your device0 could map the uart iomem as nonsecure.

yanyan-wrs replied:
accepted and fixed

etienne-lms commented on core/arch/arm/plat-imx6/imx6_init.S:
```diff
@@ -0,0 +1,345 @@
+
+	/* invalidate all cache ways */
+
+    movw r1, #0xFFFF
```
Minor indentation issue

yanyan-wrs replied:
accepted and fixed

jenswi-linaro commented on core/arch/arm/kernel/generic_entry_a32.S:
```diff
@@ -151,9 +151,21 @@ END_FUNC wait_secondary
+ */
+
+#if defined (CFG_TEE_BUILT_IN_ARGS)
+    ldr r4, =CFG_TEE_PAGEABLE_ADDR
```
I don't object to optional hardcoding of these addresses as it doesn't impact OP-TEE much, but I suspect it will cause occasional problems when they are changed. Since something has already has put stuff at those addresses, why not supply that information in registers?

jenswi-linaro replied:
This should either be refused (probably via some `.mk` file) together with `CFG_WITH_ARM_TRUSTED_FW` or `generic_entry_a64.S` needs to be updated for consistency.

yanyan-wrs replied:
Just want to avoid having to change the bootloader to pass information in R0, R2 and LR. There're existing firmware/loaders out there that are not aware of the OP-TEE's hand off protocol. Providing "built-in arguments" lowers the requirement as firmware/loader are only required "jump to" the OP-TEE Entry.
As to ARM Trusted Firmware, I think there is no need to deny the "built-in argument" when used together with it. For example, if the FW always passes the same arguments in each boot, there is no harm for OP-TEE to switch to use the built-in arguments.
As to ARMv8, I understand the concept of "built-in arguments" applies equally to ARMv8, however, for this change I do not planned to touch the ARMv8 code, because VxWorks does not support it for now and I do not have at hand any means to test the change.

jenswi-linaro replied:
The harm with "built-in argument" together with ARM-TF is the it basically puts a requirement on ARM-TF to __never__ change the arguments for that particular platform. That's a requirement which isn't needed as the arguments always are supplied any way. Please add something like:
```
ifeq($(CFG_TEE_BUILT_IN_ARGS),y)
ifeq($(CFG_WITH_ARM_TRUSTED_FW),y)
$(error error: CFG_TEE_BUILT_IN_ARGS is incompatible with CFG_WITH_ARM_TRUSTED_FW)
endif
endif
```
to `core/arch/arm/arm.mk`

By the way, this code is used on ARMv8 too, when running in Aarch32.

yanyan-wrs replied:
accepted and fixed

jenswi-linaro commented on core/arch/arm/kernel/generic_entry_a32.S:
```diff
@@ -293,6 +305,7 @@ copy_init:
 	mov	r0, r4		/* pageable part address */
 	mov	r1, r5		/* ns-entry address */
 	mov	r2, r6		/* DT address */
+
```
Please remove this empty line

yanyan-wrs replied:
accepted and fixed

jenswi-linaro commented on core/arch/arm/plat-imx6/imx6_init.S:
```diff
@@ -0,0 +1,346 @@
+	bne loop_1
+
+	/* Allow NSec to manage FIQ/Imprecise abort */
+	mrc p15, 0, r0, c1, c1, 0    /* read Secure Configuration Register */
```
Please implement and use a `read_scr` macro instead, please fix all `mrc` and `mcr` correspondingly.

jenswi-linaro replied:
`sm_smc_entry()` sets and clears `SCR_FIQ`, will that interfere with normal world managing FIQs?

yanyan-wrs replied:
Actually I had doubt on this too; I copied this from the tz_a9init.S of plat-stm. The ARM doc says the combination of FW=1 and FIQ=1 should be avoided as apparently the NS world can mask out the FIQ thus denies FIQ to reach to OP-TEE.
However, it seems the OP-TEE does not use FIQ for this platform, and VxWorks also ignores FIQ (in GIC configuration all INTs are assigned to GROUP0), so letting NS to manage FIQs will not cause any problem. And in order to keep consistency I did change this code.

jenswi-linaro commented on core/arch/arm/plat-imx6/imx6_init.S:
```diff
@@ -0,0 +1,346 @@
+    /* configure imx6 CSU */
+
+    /* first grant access of all peripherals to NS... */
+    ldr r0, =0x021c0000
```
A define for this magic address would be nice, same goes for all the other magic addresses in this file.

yanyan-wrs replied:
accepted and fixed

jenswi-linaro commented on core/arch/arm/plat-imx6/main.c:
```diff
@@ -0,0 +1,185 @@
+#define UART_REG_OFFSET_URXD        (0x0)
+#define UART_REG_OFFSET_UTXD        (0x40)
+#define UART_REG_OFFSET_USR2        (0x98)
+    #define TXFE                    (1<<14)
```
Strange indentation
Please use the `BIT` or `BIT32` macro from `<util.h>` instead.

yanyan-wrs replied:
accepted and fixed

jenswi-linaro commented on core/arch/arm/plat-imx6/main.c:
```diff
@@ -0,0 +1,185 @@
+{
+	vaddr_t va = console_base();
+
+	while ((read32(va + UART_REG_OFFSET_USR2) & TXFE) == 0)
```
Can `imx_uart_putc()` be used instead?

yanyan-wrs replied:
Seems possible as spec says the same. I'll test it on hardware.

jenswi-linaro commented on core/arch/arm/plat-imx6/main.c:
```diff
@@ -0,0 +1,185 @@
+	/* release secondary core */
+
+	val = read32(va + SRC_SCR);
+	val = val | 1 << (SRC_SCR_CORE1_ENABLE_OFFSET + (index - 1));
```
Please use the `BIT` or `BIT32` macro from `<util.h>` instead of `1 << ...`

yanyan-wrs replied:
accepted and fixed

jenswi-linaro commented on core/arch/arm/plat-imx6/platform_config.h:
```diff
@@ -0,0 +1,192 @@
+
+/* UART1, shared with NS */
+
+#define DEVICE0_PA_BASE			ROUNDDOWN(UART1_BASE, \
```
Please use the `register_phys_mem` macro instead of defining these `DEVICE*` macros

yanyan-wrs replied:
accepted and fixed

jenswi-linaro commented on core/arch/arm/sm/smc_vxworks.c:
```diff
@@ -0,0 +1,38 @@
+
+void vxworks_smc_handler(struct thread_smc_args *args)
+{
+	switch (args->a1) {
```
Wouldn't returning some error code in `args->a0` for unhandled calls be appropriate?

yanyan-wrs replied:
You're right. Although this is a big TODO, we'll still need to return some definite thing to NS requester.

jenswi-linaro commented on core/arch/arm/plat-imx6/imx6_init.S:
```diff
@@ -0,0 +1,346 @@
+    ldr r2, =0x00330033
+    str r2, [r0]
+
+    /* TODO: TZASC init */
```
Copy paste error?

yanyan-wrs replied:
No, this is actually a real TODO thing. We are going to configure TZASC to divide the DDR into NS DDR and Sec DDR, which agrees with the DDR space definition in plat-xxx/platfom_config.h. For now the DDR separation is not protected by TZASC.

jenswi-linaro commented on core/arch/arm/plat-imx6/imx6_init.S:
```diff
@@ -311,15 +311,15 @@ loop_1:
-	orr r0, r0, #0x30            /* SCR[FW]=1, SCR[AW]=1 */
-	mcr p15, 0, r0, c1, c1, 0    /* write Secure Configuration Register */
+	read_scr r0    		/* read Secure Configuration Register */
+	orr r0, r0, #0x20  	/* SCR[FW]=0, SCR[AW]=1 */
```
The comment isn't accurate as `SCR_FW` isn't cleared. Please use `SCR_AW` instead of `#0x20`.

yanyan-wrs replied:
What a mistake I made here. OK, I corrected the code to load the SCR with an arbitrary value. The FIQ is set to be trapped into FIQ mode here, and later the SM code would change it on entering and exiting Secure World.

jenswi-linaro commented on core/arch/arm/plat-imx6/imx6_init.S:
```diff
@@ -0,0 +1,353 @@
+	 * Register ICDISR[1-31] = 0xFFFFFFFF
+	 * - All external interrupts are NonSecure.
+	 */
+	ldr  r0, =(GIC_DIST_BASE + GIC_DIST_ISR1)
```
Would it be possible to initialize GIC with `gic_init()` from `main_init_gic()` instead?

yanyan-wrs replied:
Agreed. Looks like I have based my development on an out-of-data platform model; the plat-stm seems to ignore all drivers too.
Anyway, I changed the code to supply the week symbol main_init_gic(). However, I noticed the GIC driver only provides the init routine for primary core, so I have to keep the per-core GIC init code in my assembler init file.

jenswi-linaro replied:
That's a shortcoming of the generic code,  `init_secondary_helper()` should directly or indirectly do some per-core GIC init. But that can be fixed in another PR.

jenswi-linaro commented on core/arch/arm/plat-imx6/main.c:
```diff
@@ -0,0 +1,196 @@
+					  MEM_AREA_IO_SEC);
+	gicd_base = (vaddr_t)phys_to_virt(GIC_BASE + GICD_OFFSET,
+					  MEM_AREA_IO_SEC);
+	TEE_ASSERT(gicc_base && gicd_base);
```
We've removed the `TEE_ASSERT` macro. Please do:
```
        if (!gicc_base || !gicd_base)
                panic();
```
instead

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented on core/arch/arm/plat-imx6/conf.mk:
```diff
@@ -0,0 +1,20 @@
+PLATFORM_FLAVOR ?= sabrelite
+PLATFORM_FLAVOR_$(PLATFORM_FLAVOR) := y
```
This line is not needed anymore since commit 5ef74e7.

yanyan-wrs replied:
accepted and fixed

jforissier commented:
Could you please add a commit to update `.travis.yml` to build for your platform?

Also, when you squash and rebase the commits, please add the PLATFORM value to the commit description, for instance "Add the i.MX6 Sabre Lite board support (PLATFORM=imx6)".

Regarding platform and platform flavor: shouldn't VxWorks appear somewhere? Maybe PLATFORM_FLAVOR=sabrelite_vxworks? I'm just trying to anticipate what will happen if one wants to build for i.MX6 without VxWorks.

etienne-lms commented on core/arch/arm/include/sm/optee_smc_vxworks.h:
```diff
@@ -0,0 +1,35 @@
+/* OPTEE extended SMC for VxWorks */
```
could you add a notice (copyright and license) ?

yanyan-wrs replied:
accepted and fixed

etienne-lms commented on core/arch/arm/include/sm/optee_smc_vxworks.h:
```diff
@@ -0,0 +1,35 @@
+
+#endif /* OPTEE_SMC_VXWORKS_H */
+
+
```
minor: several useless empty lines that could be removed.

yanyan-wrs replied:
accepted and fixed

yanyan-wrs commented:
@jforissier We didn't introduce the VxWorks Support as a platform flavor, since for now platform flavor represents hardware variants, and we want to respect that. 
Instead, the VxWorks Support is enabled by a CFG item:
CFG_TEE_VXWORKS_SMC: include VxWorks specific SMC request handling.
Together with two other newly introduced "OS-neutral" items:
CFG_TEE_BUILT_IN_ARGS: switch to use built-in parameters for more flexible interfacing with loaders.
CFG_TEE_SECONARY_ENTRY_L1_INV: a patch to do L1 Invalidate if loader does not do it when bringing up secondary cores.
Even with CFG_TEE_VXWORKS_SMC enabled, the resultant OP-TEE can be used together with Linux (the only downside is it contains some SMC handling branches as dead code). So in fact our change does not deviate from the fact that OP-TEE is NS OS neutral.

jforissier commented on .travis.yml:
```diff
@@ -185,5 +188,9 @@ script:
+  # Freescale i.MX6 SABRE Lite
+  - $make PLATFORM=imx6-sabrelite
+
+
```
Add only one blank line here

jforissier commented on mk/config.mk:
```diff
@@ -184,3 +184,24 @@ CFG_DTB_MAX_SIZE ?= 0x10000
 CFG_TEE_CORE_EMBED_INTERNAL_TESTS ?= y
+
+# For platforms that do not invalidate L1 before entering OP-TEE's SMP entry,
+# enable this for OP-TEE to invaldiate L1 in secondary core entry in the
```
Typo

jforissier commented on mk/config.mk:
```diff
@@ -184,3 +184,24 @@ CFG_DTB_MAX_SIZE ?= 0x10000
+# VxWorks uses OP-TEE as both a TEE and the monitor mode program. This enables
+# the OS-specific SMC handling so that OP-TEE can answer to those requests.
+CFG_TEE_VXWORKS_SMC ?= n
+
```
Useless blank line

jforissier commented:
@yanyan-wrs OK thanks for clarifying, I'm fine with this.

One thing I find a bit confusing: we already have `plat-imx` with flavor `mx6ulevk`, which is identified as "FSL i.MX6 UltraLite EVK Board" in `README.md`. Now you're introducing `plat-imx6` with flavor `sabrelite` which is "i.MX6 SABRE Lite". Are these two platforms sufficiently different to be two different $PLATFORM values? (sorry I could check the code myself but you probably know the boards better than I do).

Also, could you add the new platform to `README.md`? (section 3 "Platforms supported", in the table marked  "Please keep this list sorted in alphabetic order").

There is also the `MAINTAINERS.md` file: feel free to add your name there if you're OK with being contacted occasionally for help. For instance, when we are about to make a new tagged release, we ask all maintainers to validate on there platform. Without a maintainer assigned to a platform, we do our best to keep the code maintained, but we can't test since we usually don't have the hardware. You don't have to do it if you don't want to, it's really up to you.

One last thing regarding your Signed-off-by: tag. We require people to give their real name and email. You are using `yanyan-wrs <yan.yan@windriver.com>`, which looks like some login name or pseudonym (due to the -wrs suffix). May I suggest `Yan Yan <yan.yan@windriver.com>` instead?

yanyan-wrs commented:
@jforissier The mx6ulevk is cortex-a7 while sabrelite is cortex-a9, although that does not prevent them from sharing a common plat. You're correct, and I'll try merging them together, hopefully it will not break anything in mx6ulevk since I don't have that board for testing.
I'll add myself to maintainer list and change my signed-off-by tag as you suggested.
Thanks for these valuable tips.

jforissier commented:
@yanyan-wrs thank you, your effort is appreciated.

Support for `PLATFORM=imx-mx6ulevk` was added by @MrVan who has registered as a maintainer. So, @MrVan, please let us know your opinion on merging `plat-imx` and `plat-imx6` (this PR) as suggested above.

yanyan-wrs commented:
I've made a tentative merge between plat-imx and plat-imx6 (see latest commit). The two flavors are mx6ulevk (i.MX6 UltraLite EVK) and mx6qslevk(i.MX6q SABRE Lite EVK).
As you can see, there are two kind of differences that prevents code reuse:
1. arch difference: e.g., cortex-a9 use PL310, while cortex-a7 has ARM Generic Timer.
2. the plat model is not updated for plat-imx, while plat-imx6 is forced to use the latest model, e.g., the use of register_phys_mem.

MrVan commented on core/arch/arm/plat-imx/imx6_init.S:
```diff
@@ -0,0 +1,342 @@
+	ldr r0, =CSU_CSL_16 /* TZASC1,TZASC2 */
+	ldr r2, =0x00330033
+	str r2, [r0]
+
```
@yanyan-wrs 
The CSU configuration should be per platform from my view.  Since you named this file as imx6_init.S, this file may be shared by i.MX6SDL/SX/SL and etc. I prefer move the csu configuration to a per platform file, but not shared with i.MX6 family. What do you think?

yanyan-wrs replied:
You're absolutely right. I'll change the file name to a9_plat_init.S since the core initialization code applies to all cortex-a9 based i.MX6 variants; for platform specific things such as the CSU, I'll protect them with corresponding PLATFORM_FLAVOR macro.

MrVan commented on core/arch/arm/plat-imx/platform_config.h:
```diff
@@ -114,7 +112,150 @@
+#define CSU_CSL_5				0x021c0014
+#define CSU_CSL_16				0x021c0040
+#define CONSOLE_UART_BASE		UART1_BASE
+#define CONSOLE_UART_PA_BASE	UART1_BASE
```
I think you should assign a new value for CONSOLE_UART_BASE. Virutal address <=64M is used by OPTEE kernel and OPTEE ta. 

Also check CSU and GIC address. I did not check the virual address, you need to make sure the virtual address > 0x4000000

yanyan-wrs replied:
I think this is no longer a problem. Some recent commit added support for non 1:1 mapping and all mapping is done to virtual pool above kernel space.

MrVan commented:
@jforissier I agree to use plat-imx. 
@yanyan-wrs I see you use a follow up patch to merge your patch to use plat-imx. I do not insist, but I think better to use plat-imx in your first patch which is not upstreamed, but not use a follow up patch.

jforissier commented:
@MrVan thanks for taking the time to review this.
About the follow-up patch: I think @yanyan-wrs pushed a new patch to make the review process easier, that is, to keep the commits unchanged during the review. And, I agree with you, in the end the code should use `plat-imx` directly.

MrVan commented on core/arch/arm/plat-imx/main.c:
```diff
@@ -37,11 +39,31 @@
+
 static void main_fiq(void);
+#if defined(PLATFORM_FLAVOR_mx6qsabrelite) || \
+	defined(PLATFORM_FLAVOR_mx6qsabresd)
```
Drop the if defined and endif.

yanyan-wrs replied:
accepted and fixed

MrVan commented on core/arch/arm/plat-imx/main.c:
```diff
@@ -37,11 +39,31 @@
 	.std_smc = tee_entry_std,
+#if defined(PLATFORM_FLAVOR_mx6ulevk)
 	.fast_smc = tee_entry_fast,
+#endif /* defined(PLATFORM_FLAVOR_mx6ulevk) */
```
Drop if defined. Let them all use platform_tee_entry_fast.

yanyan-wrs replied:
accepted and fixed

MrVan commented on core/arch/arm/plat-imx/main.c:
```diff
@@ -37,11 +39,31 @@
+	defined(PLATFORM_FLAVOR_mx6qsabresd)
+	.fast_smc = platform_tee_entry_fast,
+#endif
 	.fiq = main_fiq,
```
Ditto

yanyan-wrs replied:
Google says it's a Pokemon?

NightEugene replied:
Catch him!

MrVan commented on core/arch/arm/plat-imx/main.c:
```diff
@@ -97,3 +129,88 @@ void console_flush(void)
+}
+
+#ifdef CFG_TEE_VXWORKS_SMC
+
```
This piece code is not vxworks specific. Use platform ifdef here.

yanyan-wrs replied:
Correct, it's generic secondary boot up code

MrVan commented on core/arch/arm/plat-imx/a9_plat_init.S:
```diff
@@ -0,0 +1,350 @@
+
+#if defined(PLATFORM_FLAVOR_mx6qsabrelite) || \
+	defined(PLATFORM_FLAVOR_mx6qsabresd)
+
```
How about using C language for this CSU configuration?

yanyan-wrs replied:
I didn't find any hook point/weak symbol in the generic boot process where I can add C init code. Any advise here?

MrVan replied:
@yanyan-wrs Thinking add a function in main.c, named `int init_csu(void)`. Then in a9_plat_init.S, use `bl init_csu`.

yanyan-wrs replied:
I revisited the proposal and I think I'll keep the CSU configuration in ASM, since I found the logic is quite simple, and other platform/cpu init code is also kept here; most of the main.c deal with hardware that are in used during op-tee runtime, it might cause some confusion to move the one-time-config CSU code here. What do you think?

NightEugene commented:
Hi all,

Sorry for off topic. I am trying to run optee with Linux Kernel on i.MX6 Sabre Quad.
This patch very helped me. I have success when I run Linux Kernel without outer cache.
But when I run Linux Kernel with outer cache even when I don't define CFG_PL310 in optee Linux Kernel hangs on enabling MMU. Where can be my problem? Thanks a lot

etienne-lms commented:
@NightEugene the issue is, i think, that PL310 outer cache, on cortex-a9 (as imx6 sabre are), can be enable/disable (and mostly configured) only from the TZ secure state. Hence non-secure linux cannot, from it own, enable outercache. If trying so, linux will crash.

An implementation is to create a secure service in optee to configure/enable/disable outer cache, and request linux to call such service for cache enable/disable. 
Optee **does not** currently implement such service. 
Linux cache-l2x0.c provides `l2c_write_sec` handlers for such purpose.

Another implementation is to have the secure inits enabling the outer cache, then locking and invalidating the lines before linux is booted. Doing so allows linux to enable (resp. disable) outer cache by unlocking (resp. locking and invalidating) outer cache lines.
Optee proposes this implementation through the CFG_PL310_LOCKED directive.
Linux kernel cache-l2x0.c already unlocks lines when enabling outer cache.
Note that in this setup, PL310 Auxiliary Control register must be configure to allow non-secure to lock/unlock cache lines (AUXCTRL, bit#26).

NightEugene commented:
@etienne-lms, thanks for you answer!
It sounds reasonable, but Linux Kernel hangs before calling start_kernel. Hence Linux didn't try to enable PL320 outer cache. It only executed code from 'stext' and stopped on '__turn_mmu_on'.

yanyan-wrs commented:
@NightEugene Could you please provide me with the steps of how you configure Linux and disable L2 cache; I would like to investigate this issue further and make sure that this patch works for both VxWorks and Linux. Thanks.

NightEugene commented:
@yanyan-wrs,
I make imx_v6_v7_defconfig, then set:
```
CONFIG_OUTER_CACHE=n
CONFIG_OUTER_CACHE_SYNC=n
CONFIG_MIGHT_HAVE_CACHE_L2X0=n
CONFIG_CACHE_L2X0=n
```
Also I comment line in core/arch/arm/plat-imx/conf.mk:
`#$(call force,CFG_PL310,y)`

Moreover I added delay in Linux arch/arm/kernel/head.S:
```
ENTRY(__turn_mmu_on)
#if 1 // my delay
    ldr r7, = 0xffffff
loop:
    sub r7, #1
    cmp r7, #0
    bne loop
#endif
    mov r0, r0
    instr_sync
    mcr p15, 0, r0, c1, c0, 0       @ write control reg 
    mrc p15, 0, r3, c0, c0, 0       @ read id reg 
    instr_sync
    mov r3, r3
    mov r3, r13 
    ret r3
__turn_mmu_on_end:
ENDPROC(__turn_mmu_on)
```
With this delay Linux booted, but has many problems like
```
[    1.009613] CPU1: failed to come online
[    2.009538] CPU2: failed to come online
[    3.009538] CPU3: failed to come online
```
I still cannot understand why this delay helps, but without it Linux hangs on __turn_mmu_on.

yanyan-wrs commented:
@NightEugene 
I tried Linux, it hangs in __fixup_pv_table() in head.S, whether PL310 is turn on or off.

Following is how I did it:
1. Build OP-TEE with:
```
CFG_TEE_BUILT_IN_ARGS=y
CFG_TEE_NS_ENTRY_ADRS=0x12000000
CFG_TEE_GDB_BOOT=y
```
0x12000000 is the place I'll put the Linux zImage. CFG_TEE_GDB_BOOT ensure OP-TEE relays r0-r2 to Linux.

2. Use mkimage to produce the uTee image. Make it looks like uImage so the bootm command can boot it.
`mkimage -A arm -O linux -C none -a 0x4dffffe4 -e 0x4e000000 -d tee.bin uTee`

3. Make the linux kernel use imx_v6_v7_defconfig.

4. In u-boot kernel, load everything in place and use bootm to enter OP-TEE:
```
tftp 0x12000000 zImage
tftp 0x18000000 imx6q-sabresd.dtb 
tftp 0x20000000 uTee
setenv bootargs console=ttymxc0,115200 root=/dev/mmcblk0p2 rootwait rw
bootm 0x20000000 - 0x18000000
```

I checked the the execution can enter the NS zImage at 0x12000000, and r0-r2 inputs are correct. However, the execution enters the __fixup_pv_table and never get out. It appears it's trapped infinitely in the loop in function __fixup_a_pv_table. I'll investigate further to see why.

Still some general questions:
1. Can Linux tell it's in NS and refrain itself from accessing privileged resources? Or should something be added in CONFIG?
2. Should the DT node for "/tee" device be manually added? and the same with the memory description, should the Shared Memory be manually subtracted from DT node? I noticed there is an OP-TEE option CFG_DT which will modify DT accordingly during runtime, shall we rely on that?
3. Any general suggestion/description as to how to run Linux and OP-TEE together. @etienne-lms your description of L2's behavior in NS world is highly helpful, really appreciate it.

NightEugene commented:
@yanyan-wrs, thank you for researching this issue

Your method making uTee image are cool! I would like use it :)

> 2 . Should the DT node for "/tee" device be manually added? and the same with the memory description, should the Shared Memory be manually subtracted from DT node? I noticed there is an OP-TEE option CFG_DT which will modify DT accordingly during runtime, shall we rely on that?

Now I use CFG_DT option and it looks like OP-TEE adds DT node for /tee and memory description correctly. During Linux booting I got
```
[    8.501425] optee firmware:optee: probing for conduit method from DT.
[    8.509727] optee firmware:optee: initialized driver
```
And tee-supplicant works

yanyan-wrs commented:
@NightEugene 
Ok, I googled and it says that certain U-BOOT version leaves L2 cache on, which hangs the kernel if the CONFIG_ARM_PATCH_PHYS_VIRT is enabled:
[https://groups.google.com/forum/#!topic/linux.kernel/x7ULaIibC8o](url)
So I disable L2 in OP-TEE by adding CFG_PL310_LOCK (I think you cannot disable L2 by just commenting out CFG_PL310, OP-TEE just ignores it, but does not shut it down; while CFG_PL310_LOCK locks down the L2 and clears all contents), then I can reach the place before turning MMU on in SCTLR. However, the kernel still die after that, and I don't have a HW debugger so it becomes hard to trace the execution after MMU on (I am playing with the LL debug option, but no luck right now).

WRT the secondary core boot, I think it is foreseeable that they cannot boot-up, because the platform code should be patched to issue certain SMC request for OP-TEE to bring these core up and put them through the OP-TEE init and go to NS entry. The VxWorks has supported this logic, but Linux is not patched yet.

etienne-lms commented:
> (...) platform code should be patched to issue certain SMC request for OP-TEE
> to bring these core up and put them through the OP-TEE init and go to NS entry.
> The VxWorks has supported this logic, but Linux is not patched yet.

actually this is what pcsi is about. maybe some day optee will support psci for armv7 :)
(on armv8: psci is provided by the EL3 (i.e arm-trusted-firmware), outside of optee).

yanyan-wrs commented:
I just found the imx_init_l2cache() function in arch/arm/mach-imx/system.c has accessed PL310 control registers in Non-Secure world, which hangs the kernel. This is the case in the latest 4.7.2 kernel as well as Freescale's imx linux repository.
@NightEugene: since you managed to get over it, do you use a different version of Linux?

NightEugene commented:
@yanyan-wrs,
I use 4.5.0 kernel from linaro-swg repository (optee branch). In any case I compile kernel without L2 cache feature. You reached start_kernel?

yanyan-wrs commented:
@NightEugene 
After I locked (i.e., disabled) L2 in OP-TEE, and comment out the L2 access in imx_init_l2cache() in Linux, then UP kernel works fine; SMP is not OK; secondary cores are not up, as expected, because no SMC is sending to OP-TEE to release secondary cores.

yanyan-wrs commented:
One more thing, the original imx_enable_cpu() function accesses the SRC(system reset control) to bring up secondary cores. If works with OP-TEE, SRC is kept away from NS world for security reasons, so the normal access to SRC should cause exceptions, I wonder why in your setup, the execution can pass the smp_init() and just report some error of "failed to come online".

NightEugene commented:
@yanyan-wrs 
About L2 access you are right, there are same issue https://community.nxp.com/thread/379983.
We can check if l2x0 controller is already enabled.
```
/* Check if l2x0 controller is already enabled.
* If you are booting from non-secure mode
* accessing the below registers will fault.
*/
if (!(readl_relaxed(l2x0_base + L2X0_CTRL) & L2X0_CTRL_EN)) {
  // do the rest
}
```

About why I have "failed to come online" I still don't understand.
I think we can up secondary cores from u-boot with 
```
void smp_set_core_boot_addr(unsigned long addr, int corenr)
{
    int cpu, val;

    for (cpu = 1; cpu <= 3;  ++cpu) {
        writel(addr, SRC_BASE_ADDR + 0x20 + 8 * cpu);
        writel(0, SRC_BASE_ADDR + 0x20 + 8 * cpu + 4);
        val = readl(SRC_BASE_ADDR);
        val |= 1 << (SRC_SCR_CORE_1_RESET_OFFSET + cpu - 1);
        val |= 1 << (SRC_SCR_CORE_1_ENABLE_OFFSET + cpu  -1);
        writel(val, SRC_BASE_ADDR);
    }
}
```
And call it before kernel_entry
```
smp_set_core_boot_addr((unsigned int)0x4e000000, -1)
kernel_entry(0, machid, r2);
```
Also we use CFG_BOOT_SYNC_CPU option in OP-TEE. But it doesn't work(
I hope that soon I will have JTAG and it will help to understand where is problem.

Could you tell where did you find 4.7.2 kernel? I see only 4.6.0 here https://github.com/Freescale/linux-fslc

yanyan-wrs commented:
@NightEugene 
I updated a new version, which supports Linux SMP. Now I have removed the vxWorks specific SMC requests to make it an generic secondary core boot request for any Rich OS.

I tested with both SabreLite and SabreSD board, everything is fine. 
Following is how to do it:
1. Update the latest git code, and build OP-TEE with following confgurations:
```
CFG_BUILT_IN_ARGS=y 
CFG_PAGEABLE_ADDR=0 
CFG_NS_ENTRY_ADDR=0x12000000 
CFG_DT_ADDR=0x18000000 
CFG_TEE_GDB_BOOT=y 
CFG_DT=y 
CFG_BOOT_SECONDARY_REQUEST=y 
PLATFORM=imx-mx6qsabrelite
```
2. Pretend the tee as an Linux uImage. This is to let u-boot pass r0-r2 as if it is booting a Linux, and with the CFG_TEE_GDB_BOOT option above, OP-TEE will pass r0-r3 to the real Rich OS Linux.
```
mkimage -A arm -O linux -C none -a 0x4dffffe4 -e 0x4e000000 -d tee.bin uTee
```
3. Modify the SMP boot code of Linux, in linux/arch/arm/mach-imx/platsmp.c:
add extra header and definitions:
```
#include <linux/arm-smccc.h>

#define OPTEE_SMC_CALLID_BOOT_SECONDARY       12
#define OPTEE_SMC_BOOT_SECONDARY  \
    ARM_SMCCC_CALL_VAL(ARM_SMCCC_FAST_CALL, ARM_SMCCC_SMC_32, \
                ARM_SMCCC_OWNER_TRUSTED_OS, OPTEE_SMC_CALLID_BOOT_SECONDARY)
```

change the `imx_boot_secondary`. NOTE: the original code set the entry as `v7_secondary_startup`, which has made an access to an diagnostic co processor, but it is not accessible from NS, so we just entry `secondary_startup` instead of `v7_secondary_startup`. 
```
static int imx_boot_secondary(unsigned int cpu, struct task_struct *idle)
{
#if 0
	imx_set_cpu_jump(cpu, v7_secondary_startup);
	imx_enable_cpu(cpu, true);
#endif
#if 1
    struct arm_smccc_res  res;
    arm_smccc_smc(OPTEE_SMC_BOOT_SECONDARY,
        cpu, virt_to_phys(secondary_startup),
        0, 0, 0, 0, 0, &res);
#endif
return 0;
}
```
5. You may also need to change the NS L2 initialization code, if not fixed already, as you pointed out before.

yanyan-wrs commented:
Hi everyone, sorry for the repeated changes here. I had intended to support VxWorks only, however, with help from @NightEugene, Linux Rich OS is also supported, with the existing u-boot setup and some patches to the Linux kernel regarding Non-Secure executing.
To support both Linux and VxWorks, I added a generic OP-TEE Trusted OS SMC Request:
```
/*
 * Release of secondary cores
 *
 * OP-TEE in secure world is in charge of the release process of secondary
 * cores. The Rich OS issue the this request to ask OP-TEE to boot up the
 * secondary cores, go through the OP-TEE per-core initialization, and then
 * switch to the Non-secure world with the Rich OS provided entry address.
 *
 * Call register usage:
 * a0	SMC Function ID, OPTEE_SMC_BOOT_SECONDARY
 * a1   smp cpu index
 * a2   NS entry address low
 * a3   NS entry address high
 * a4-7	Not used
 *
 * Normal return register usage:
 * a0	OPTEE_SMC_RETURN_OK
 * a1-7	Preserved
 *
 * Not idle return register usage:
 * a0	OPTEE_SMC_RETURN_EBUSY
 * a1-7	Preserved
 */
#define OPTEE_SMC_FUNCID_BOOT_SECONDARY  12
#define OPTEE_SMC_BOOT_SECONDARY \
	OPTEE_SMC_FAST_CALL_VAL(OPTEE_SMC_FUNCID_BOOT_SECONDARY)
```
With this added request, both Linux and VxWorks Rich OSes can ask for OP-TEE to release the secondary cores thus enter SMP execution.
This change is final from my part, please review it.

jenswi-linaro commented on core/arch/arm/include/sm/optee_smc.h:
```diff
@@ -353,6 +352,33 @@
+ * Call register usage:
+ * a0	SMC Function ID, OPTEE_SMC_BOOT_SECONDARY
+ * a1   smp cpu index
+ * a2   NS entry address low
```
To harmonize better with the other descriptions please rephrase as:
```
a2 Upper 32-bits of 64-bit NS entry address
a3 Lower 32-bits of 64-bit NS entry address
```
Any particular reason the lower bit was first?
Please add a comment about what can be expected of CPSR also.

jenswi-linaro commented on core/arch/arm/plat-imx/main.c:
```diff
@@ -97,3 +120,91 @@ void console_flush(void)
+		(paddr_t)&ns_entry_addrs[index],
+		sizeof(uint32_t));
+
+	/* boot secondary cores from OP-TEE load addrss */
```
`s/addrss/address/`

jenswi-linaro commented on core/arch/arm/plat-imx/main.c:
```diff
@@ -97,3 +120,91 @@ void console_flush(void)
+	return SRC_BASE;
+}
+
+static int platform_smp_boot(int index, uint32_t entry)
```
`s/int index/size_t core_idx/`

jenswi-linaro commented on core/arch/arm/include/sm/optee_smc.h:
```diff
@@ -353,6 +352,33 @@
+ * Not idle return register usage:
+ * a0	OPTEE_SMC_RETURN_EBUSY
+ * a1-7	Preserved
+ */
```
Description of possible return value `OPTEE_SMC_RETURN_EBADCMD` missing.

NightEugene commented:
@yanyan-wrs 
Thanks very much for your help! Now everything works fine. Only CAAM (Cryptographic Acceleration and Assurance Module) doesn't work, but this is a separate issue.

For my i.MX6Q SabreAuto board I maked few changes:
```
#define UART3_BASE              0x021F0000

#define CONSOLE_UART_BASE       UART3_BASE
#define CONSOLE_UART_PA_BASE    UART3_BASE

#define DRAM0_BASE              0x10000000
#define DRAM0_SIZE              0x80000000

#define CFG_DDR_TEETZ_RESERVED_START    0x8E000000
#define CFG_DDR_TEETZ_RESERVED_SIZE     0x02000000
```

jenswi-linaro commented on core/arch/arm/kernel/generic_entry_a32.S:
```diff
@@ -432,7 +445,15 @@ UNWIND(	.cantunwind)
+	/* get ns entry from ns_entry_addrs array */
+	bl  get_core_pos
+	ldr r1, =ns_entry_addrs
+	add r1, r1, r0, lsl #2	/* paddr_t is 4 bytes in 32-bit system */
```
The size of `paddr_t` is not guaranteed to be 4 bytes on a 32-bit system. It would be better to use the `uint32_t` type instead for `ns_entry_addrs[]` as anything larger than that doesn't make sense on a 32-bit system.

jenswi-linaro replied:
Please update this comment too.

jforissier commented on .travis.yml:
```diff
@@ -166,6 +168,10 @@ script:
 
+  # i.MX6Quad SABRE
+  - $make PLATFORM=imx-mx6qsabrelite
+  - $make PLATFORM=imx-mx6qsabresd CFG_BOOT_SECONDARY_REQUEST=y
```
Should we have `CFG_BOOT_SECONDARY_REQUEST ?= y` set in `plat-imx/conf.mk` for this platform flavor? My goal is to have the most likely configuration built by default, as a general rule.

jenswi-linaro commented:
I'm OK with this now. @MrVan, do you have more comments?

MrVan commented:
@jenswi-linaro Sorry for late reply. I am ok with it.

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

MrVan commented:
Reviewed-by: Peng Fan <peng.fan@nxp.com>

yanyan-wrs commented:
Should I rebase and squash the changes before merge?

jforissier commented:
@yanyan-wrs yes, please squash the commits, add the 'Reviewed-by:' tags, and rebase on top of optee_os master. Thanks!

jforissier commented:
@yanyan-wrs ping! I can tag the commits and rebase + merge myself, if you prefer, but then the PR will show up as "closed" rather than "merged" which is not so good.

yanyan-wrs commented:
Sorry for the delay, I encountered some issue in the final testing (turned out to be issue #1071). Now the PR is ready.

jforissier commented:
@yanyan-wrs @MrVan thanks for your contribution.
